HDDS-11205. Implement a search feature for users to locate keys pending Deletion within the OM Deleted Keys Insights section

ArafatKhan2198 opened the pull request:
## What changes were proposed in this pull request?

This pull request enhances the search functionality for deleted pending keys in the Deleted Keys Insights section. The previous implementation was limited to a subset of the available data returned by the API. This update ensures accurate retrieval of specific key information by improving pagination and handling of both prefix-based searches and general search scenarios.

The patch implements the ability to search for deleted keys using a specified prefix, efficiently returning results that match the prefix across both FSO and Object Store layouts. The logic has been refactored to the OMDBInsightEndpoint, unifying the method for handling all cases where the startPrefix is provided or not.
## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-11205

## How was this patch tested?

- Unit tests were created and updated to cover various scenarios, including pagination, searches with and without startPrefix, and for both FSO and non-FSO layouts.
- Manual testing was conducted to ensure that the search functionality works as expected across different scenarios.
- We can run the following command on docker container running OM.
```
// Use the Freon tool to create 5 directories with 10 keys each 
for i in {1..5}; do
  ozone freon dfsg --path=ofs://om/s3v/fso-bucket/ -s=10000 -n=10
done

// Delete all the directories under the fso-bucket which will result in records being added to deletedTable 
ozone fs -rm -r -skipTrash ofs://om/s3v/fso-bucket
```

- Using the given API endpoint which will list out all the deleted keys under the bucket directory `1inkjkisd` which is under `fso-bucket`.
```
http://localhost:9888/api/v1/keys/deletePending?startPrefix=/s3v/fso-bucket/1inkjkisd
```
<img width="1004" alt="image" src="https://github.com/user-attachments/assets/567bdc5a-7b2e-4a25-8eab-3cb6397911ab">

ArafatKhan2198 commented:
@devmadhuu @dombizita Please take a look!

errose28 commented:
For FSO buckets, how does the search work if the parent directories or even the volume and bucket have been deleted already? The key may still be pending deletion but the prefix will have no matches once the parents are cleared.

dombizita commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/OMDBInsightSearchEndpoint.java:
```diff
@@ -325,46 +332,108 @@ public String convertToObjectPath(String prevKeyPrefix) throws IOException {
-        keyIter.seek(startPrefix);
+  @GET
+  @Path("/deletePending/search")
+  public Response searchDeletedKeys(
```
The beginning and ending of this method have a lot in common with `searchOpenKeys`. 
https://github.com/apache/ozone/blob/350a340277dbb94f7a3157594d1b705f33c1d195/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/OMDBInsightSearchEndpoint.java#L110-L117
Can we extract them? Do you think it would make sense?

dombizita reviewed (commented):
Thanks for working on this @ArafatKhan2198, overall it looks good to me, I had one idea, let me know what you think of it. Also could you answer to @errose28's comment?

ArafatKhan2198 commented:
Thanks for the comment @errose28 

The DeletedTable has a structure similar to that of the key table, following this pattern:
```
volumeName/bucketName/directoryName/KeyName/ObjectID
```

When a key is deleted and placed inside the `DeletedTable`, it is stored as a key-value pair where:

- The **key** follows the structure shown above, since the entire path is being mentioned as the key we can use that to our advantage.
- The **value** is a `RepeatedOmKeyInfo` object.

The search operation leverages the `.seek()` method provided by RocksDB tables. Users provide a prefix path (complete or partial), and the seek operator fetches the first record that matches this prefix in the `DeletedTable`. It then iterates and collects all subsequent records matching this prefix, returning them in the appropriate format. 

Even if the bucket or directory is deleted, it does not impact the search because we are only seeking values in the `DeletedTable` that match the given prefix.

Additionally, I have verified how each record key is stored in the `DeletedTable`. For example: ➖

<img width="1370" alt="image" src="https://github.com/user-attachments/assets/983359e4-6e96-48a5-b819-cf5c63bf87ca">

For example, if the user provides a search prefix like volume/bucket1/dir1/, the following record shown in the image, along with any other record that matches this prefix, will be collected and returned.

ArafatKhan2198 commented:
@dombizita @devmadhuu @errose28 
Could you please take a look.

devmadhuu commented:
Pls fix the checkstyle issue.

devmadhuu reviewed (approved):
Thanks @ArafatKhan2198 for improving the patch. Changes LGTM +1. Just fix your checkstyle issue.

devabhishekpal reviewed (approved):
Thanks @ArafatKhan2198 for the changes, looks good to me.

sumitagrawl commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/OMDBInsightSearchEndpoint.java:
```diff
@@ -325,46 +323,93 @@ public String convertToObjectPath(String prevKeyPrefix) throws IOException {
-        // If no previous key is provided, start from the search prefix.
-        keyIter.seek(startPrefix);
+  @GET
+  @Path("/deletePending/search")
```
This API is same as already existing one, @Path("/deletePending")
And prevKey Can act as Search Key for starting point of search, as rocksdb support this. Just having small change, can avoid exposing new api.

ArafatKhan2198 replied:
I agree with you, @sumitagrawl, that the `prevKey` prefix and the `searchKeyPrefix` both serve the purpose of seeking to the key provided by the user. However, in our case, the `prevKey` is specifically intended to assist with pagination. As you can see, the `searchDeletedKeys` method that returns search results for deletedKeys also includes a `prevKey` parameter, which helps paginate the results returned by the search API.

The reason we can't use prevKey as a searchPrefix of ("/deletePending") is that prevKey is designed to return records starting from a given key, regardless of any prefix. It fetches all records after the prevKey until the specified limit is reached, regardless of their prefix.

On the other hand, the `searchPrefix` works by seeking the first record that matches the given prefix and only returns records that match this prefix.

If we were to use `prevKey` as a `searchPrefix`, it would break pagination. Even if we added additional checks to distinguish between `searchPrefix` and `prevKey`, it would complicate the code, making it harder to read and maintain.

Furthermore, we’ve already implemented a search feature for /openKeys in the same class, which required significant code changes. For that reason, we decided to create a separate class, `OMDBInsightSearchEndpoint`, to handle the search for /deletedPendingKeys, which is the focus of this PR.

sumitagrawl replied:
it seems /openkeys also duplicated, we can track to fix it.

End-point functionality can not be duplicated by just purpose. If its required to have separate api, then implementation must be one only, call internally same.

prevKey search for first record (with additional case of skip that matching record). This is used in pagination context.
searchkey search for first record to be returned (this search can be partial also). But both can not be used together.

The new api have extra parameter of searchKey which is optional (can not use both searchkey and prvKey together).
So either discard previous api to use new one. Or extend previous api to include searchKey.

ArafatKhan2198 commented:
Thanks @devabhishekpal @devmadhuu @sumitagrawl for the review and comments on this PR please consider the latest changes :-

ArafatKhan2198 commented:
Thanks for the review @devmadhuu @sumitagrawl @devabhishekpal please consider the latest commit having the final changes as discussed offline 

In the latest commit, we made significant improvements to the search functionality for deleted keys:

- Unified Search Logic: The search logic has been refactored and moved to OMDBInsightEndpoint. This ensures that both cases (with and without startPrefix) are handled by the same method, simplifying the code.

- Pagination Handling: We refined the logic to ensure accurate pagination in all scenarios. Now, searches correctly handle the `prevKey` and return the appropriate records, whether a `startPrefix` is provided or not.

- Expanded Unit Testing: Added new unit tests in TestOmDBInsightEndPoint and TestDeletedKeysSearchEndpoint to cover the four key search scenarios:
  -   `prevKey` provided, `startPrefix` empty.
  -   `prevKey` empty, `startPrefix` empty.
  -   `startPrefix` provided, `prevKey` empty.
  -   `startPrefix` provided, `prevKey` provided.

sumitagrawl commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/OMDBInsightEndpoint.java:
```diff
@@ -476,17 +426,159 @@ public Response getDeletedKeySummary() {
+   * If limit is 0, all matching keys are retrieved. If both startPrefix and prevKey are empty, the method starts
+   * from the beginning of the table.
+   */
+  public static <T> Map<String, T> retrieveKeysFromTable(
```
retrieveKeysFromTable  Code can be make similar to one added in ReconUtils.java to avoid duplidates

sumitagrawl reviewed (approved):
LGTM

devmadhuu commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/ReconUtils.java:
```diff
@@ -596,6 +600,70 @@ public static long convertToEpochMillis(String dateString, String dateFormat, Ti
+      // This is implicit in the following loop, as the iterator will start from the beginning
+
+      // Iterate through the keys while adhering to the limit (if the limit is not zero)
+      while (keyIter.hasNext() && (limit == 0 || matchedKeys.size() < limit)) {
```
IMO, for `limit == 0`, number of records returned should be zero irrespective of startPrefix and prevKey...We should keep consistency across ozone for limit meaning. For limit == -1 , we should return all records.

ArafatKhan2198 replied:
Thanks taken care of this inside `extractKeysFromTable` method of ReconUtils.

devmadhuu commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/OMDBInsightEndpoint.java:
```diff
@@ -476,17 +427,95 @@ public Response getDeletedKeySummary() {
+          return createBadRequestResponse("Invalid startPrefix: Path must be at the bucket level or deeper.");
+        }
+        keysFound = getPendingForDeletionKeyInfo(limit, prevKey, startPrefix, deletedKeyInsightInfo);
+      } else {
```
We can do some optimization here. We may not need else here and just pass startPrefix as it is after validation if startPrefix is valid or not. Because whether startPrefix is empty or not, those cases are being handled in `extractKeysFromTable` method already.

ArafatKhan2198 replied:
I agree, have made the changes.

devmadhuu commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/OMDBInsightSearchEndpoint.java:
```diff
@@ -386,4 +343,20 @@ private KeyEntityInfo createKeyEntityInfoFromOmKeyInfo(String dbKey,
     return keyEntityInfo;
   }
 
+  private boolean validateStartPrefixAndLimit(String startPrefix) {
```
1. We are not validating limit here as method name says.
2. Similar method is duplicated in OMDBInsightEndPoint class also. Can we optimize here ?

ArafatKhan2198 replied:
Made changes.

devmadhuu reviewed (commented):
Thanks @ArafatKhan2198 for improving the patch. Few comments left.

ArafatKhan2198 commented:
Please take another look @devmadhuu

adoroszlai commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/ReconUtils.java:
```diff
@@ -596,6 +600,88 @@ public static long convertToEpochMillis(String dateString, String dateFormat, Ti
+   *  - Seeks to prevKey, skips it, and returns subsequent keys that match startPrefix, up to the limit.
+   *
+   *  This method also handles the following limit scenarios:
+   *  - If limit = 0 or limit < -1, no records are returned.
+   *  - If limit = -1, all records are returned.
+   *  - For positive limits, it retrieves records up to the specified limit.
```
Use `{@code limit < -1}` to avoid javadoc failure.  Also consider using `{@code limit == 0}` and `{@code limit == -1}` for consistency.

Javadoc text is treated as HTML, and characters like `<` and `>` have special meaning (start/end of HTML tag).  Therefore they need to be escaped when used as non-tag.

For the same reason, plain text formatting (e.g. empty lines for separating paragraphs, `-` for list items) is lost when converted to HTML.  Consider using HTML formatting.

You may quickly check the rendered version in IDEA: mouse over the method name, javadoc is displayed in tooltip.

ArafatKhan2198 replied:
Thanks for the help @adoroszlai it worked!

devmadhuu commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/OMDBInsightEndpoint.java:
```diff
@@ -477,17 +428,90 @@ public Response getDeletedKeySummary() {
+      LOG.debug("Invalid startPrefix provided: {}", startPrefix, e);
+      return createBadRequestResponse("Invalid startPrefix: " + e.getMessage());
+    } catch (IOException e) {
+      LOG.debug("I/O error while searching deleted keys in OM DB", e);
```
Since we are wrapping only `e.getMessage` in returned response, so exact issue may be difficult to know in case of real time cluster when we have used `LOG.debug`. Is it not better to use `LOG.error` because once some issue occurs in real time cluster, we again try to repro by enabling debug level. Same comment for other catch blocks.

ArafatKhan2198 replied:
I agree the JSON response would not be giving much hint in further debugging, changed it to ERROR
Thanks.

devmadhuu commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/OMDBInsightSearchEndpoint.java:
```diff
@@ -389,4 +346,20 @@ private KeyEntityInfo createKeyEntityInfoFromOmKeyInfo(String dbKey,
     return keyEntityInfo;
   }
 
+  private boolean validateStartPrefix(String startPrefix) {
```
This similar method is duplicated in `OMDBInsightEndpoint` class also. Given this comment earlier but not fixed.

ArafatKhan2198 replied:
Done

devmadhuu reviewed (commented):
Thanks @ArafatKhan2198 for improving the patch. Just few nits and one old comment not yet addressed.

devmadhuu commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/OMDBInsightSearchEndpoint.java:
```diff
@@ -110,25 +116,17 @@ public OMDBInsightSearchEndpoint(OzoneStorageContainerManager reconSCM,
-        return createBadRequestResponse(
-            "Invalid startPrefix: Path must be at the bucket level or deeper.");
+      // Validate the request parameters
+      if (!validateStartPrefix(startPrefix)) {
```
Here, we are not checking if startPrefix is blank or not ? Any reason ?

devmadhuu reviewed (commented):
One minor comment.

devmadhuu commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/OMDBInsightEndpoint.java:
```diff
@@ -477,17 +429,90 @@ public Response getDeletedKeySummary() {
+
+    try {
+      // Validate startPrefix if it's provided
+      if (isNotBlank(startPrefix) && !validateStartPrefix(startPrefix)) {
```
Then why it is needed to be checked separately here for blank value of `startPrefix` ?

ArafatKhan2198 replied:
The reason for adding a blank check along with validateStartPrefix is tied to the default value I've assigned to startPrefix as an empty string ("") for the /deletePending search. This choice is because both prevKey and startPrefix have similar roles, but they are used differently. There are four key cases where either prevKey or startPrefix can be empty or hold a value, and assigning an empty string ("") to startPrefix—instead of "/"—helps handle these cases more effectively.

The four cases are:

- prevKey provided, startPrefix empty:
 In this case, we seek to the prevKey, skip it, and return the subsequent records up to the limit.

- prevKey empty, startPrefix empty:
 Here, we iterate from the start of the table and retrieve all records up to the limit.

- startPrefix provided, prevKey empty:
In this scenario, we seek to the first key that matches the startPrefix and return all matching keys up to the limit.
 
- startPrefix provided, prevKey provided:
 In this case, we seek to the prevKey, skip it, and return the subsequent keys that match the startPrefix, up to the limit.

By setting the default startPrefix as an empty string, it simplifies handling these cases, especially in the forthcoming PR where I'll be integrating open keys search in omDBInsightEndpoint. This ensures that prevKey and startPrefix work smoothly together, making the logic easier to follow.

devmadhuu replied:
@ArafatKhan2198 I understand that why we are assigning the `startPrefix` as empty. My original question was not about why you are assigning the `startPrefix` as empty. Rather the point is why there is a. need to validate that if API is already setting default as EMPTY String. That check is redundant.

devmadhuu reviewed (commented):
Pls check , some inconsistency for blank value check of `startPrefix` at some places but not a some places.

devmadhuu reviewed (commented):
Pls check my comment.

devmadhuu commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/ReconUtils.java:
```diff
@@ -596,6 +600,103 @@ public static long convertToEpochMillis(String dateString, String dateFormat, Ti
+    // If limit = -1, set it to Integer.MAX_VALUE to return all records
+    int actualLimit = (limit == -1) ? Integer.MAX_VALUE : limit;
+
+    try (TableIterator<String, ? extends Table.KeyValue<String, T>> keyIter = table.iterator()) {
```
Need to do null check for table, as we have faced issues in past when `omMetaManager` is still initalizing and may throw NPE here on `table` object

devmadhuu reviewed (commented):
Some minor comment.

ArafatKhan2198 commented:
> @ArafatKhan2198 I understand that why we are assigning the startPrefix as empty. My original question was not about why you are assigning the startPrefix as empty. Rather the point is why there is a. need to validate that if API is already setting default as EMPTY String. That check is redundant.

@devmadhuu 
The reason for adding a check to see if the startPrefix is default (empty) is to handle cases where both startPrefix and prevKey are empty. In such scenarios, we need to iterate through all the records in the table.

Without this null check, validateStartPrefix would return false because it mistakenly interprets the empty startPrefix as a volume-level search, which isn't allowed. By adding a check for an empty string, we explicitly allow this condition where startPrefix is `""`. Without this, for the default value of `""`, we would encounter an error and the following would be logged:
```
return createBadRequestResponse("Invalid startPrefix: Path must be at the bucket level or deeper.");
```
This adjustment ensures the search behaves correctly when startPrefix is empty, avoiding unnecessary errors.

devmadhuu reviewed (approved):
Thanks @ArafatKhan2198 for improving the patch. Changes LGTM +1. Kindly check that API 404 error when we landup on the keyPendingForDeletion page in Recon UI. And I understand that for openKeys search API, you will be fixing things in follow up PR. Rest all looks good.

devmadhuu reviewed (approved):
Changes LGTM now. Thanks @ArafatKhan2198
